transformations/chinese_butter_fingers_perturbation/transformation.py

```diff
@@ -0,0 +1,80 @@
+import itertools
+import random
+import unidecode as u
+import json
+import os
+
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+def butter_finger(text, prob=0.2, keyboard="pinyin", seed=0, max_outputs=1):
+    random.seed(seed)
+
+    prob_of_typo = prob * 100
+    perturbed_texts = []
+    for _ in itertools.repeat(None, max_outputs):
+        butter_text = ""
+        for chinese_character in text:
+            similar_pinyins = get_characters_with_similar_pinyin(chinese_character, keyboard)
+
+            if random.choice(range(0, 100)) <= prob_of_typo and similar_pinyins != "":
+                new_chinese_character = random.choice(similar_pinyins)
+            else:
+                new_chinese_character = chinese_character
+
+            butter_text += new_chinese_character
+        perturbed_texts.append(butter_text)
+    return perturbed_texts
+
+def get_characters_with_similar_pinyin(chinese_character, keyboard):
+    chinese_character_database = load_chinese_character_data_into_memory()
+
+    pinyin_for_char_to_be_perturbed = ""
+    if keyboard == "pinyin":
```

Comment:
If the variable `keyboard` only allows one value, it is not necessary to have an if-clause here, and the `keyboard` can be removed as well.

Reply:
I had an initial idea about adding the capability to deal with keyboards that inputs Pinyin with tone marks (e.g wǒ ài nǐ) and convert them to Chinese characters (e.g 我爱你) then perturbing them and possibly changing them back. 

Something like this keyboard input method:
https://mandarinhq.com/2020/05/type-pinyin-tone-marks/

Although I believe those sort of text is quite uncommon compared to Chinese characters so I am not sure whether this would be useful or not. I could see the conversion from pinyin tones to Chinese characters being a useful transformation.

Reply:
The transformation based on Pinyin and tones can be a more constrained version of the current one. I think it would be more natural as it also embeds some phonetic similarities (some people might use voice input while typing). If the tone one is not difficult to implement, I think you could keep both in this transformation. Maybe adding an argument in the `__init__()` function like `consider_tone=False` to limit whether the generated word should have the same Pinyin **and** tone as the original word.

Reply:
Thanks for the great idea! I have implemented this as per your suggestion.